guppylang/diagnostic.py

```diff
@@ -109,3 +151,222 @@ class Error(Diagnostic, Protocol):
     compilation."""
 
     level: ClassVar[Literal[DiagnosticLevel.ERROR]] = DiagnosticLevel.ERROR
+
+
+@runtime_checkable
+@dataclass(frozen=True)
+class Note(SubDiagnostic, Protocol):
+    """Compiler sub-diagnostic giving some additional context."""
+
+    level: ClassVar[Literal[DiagnosticLevel.NOTE]] = DiagnosticLevel.NOTE
+
+
+@runtime_checkable
+@dataclass(frozen=True)
+class Help(SubDiagnostic, Protocol):
+    """Compiler sub-diagnostic suggesting how to fix something."""
+
+    level: ClassVar[Literal[DiagnosticLevel.HELP]] = DiagnosticLevel.HELP
+
+
+class DiagnosticsRenderer:
+    """Standard renderer for compiler diagnostics in human-readable format."""
+
+    source: SourceMap
+    buffer: list[str]
+
+    #: Maximum amount of leading whitespace until we start trimming it
+    MAX_LEADING_WHITESPACE: Final[int] = 12
+
+    #: Amount of leading whitespace left after trimming for padding
+    OPTIMAL_LEADING_WHITESPACE: Final[int] = 4
+
+    #: Maximum length of span labels after which we insert a newline
+    MAX_LABEL_LINE_LEN: Final[int] = 60
+
+    #: Maximum length of span labels after which we insert a newline
+    MAX_MESSAGE_LINE_LEN: Final[int] = 80
+
+    #: Number of preceding source lines we show to give additional context
+    PREFIX_CONTEXT_LINES: Final[int] = 2
+
+    def __init__(self, source: SourceMap) -> None:
+        self.buffer = []
+        self.source = source
+
+    def render_diagnostic(self, diag: Diagnostic) -> None:
+        """Renders a single diagnostic together with its sub-diagnostics.
+
+        Example:
+
+        ```
+        Error: Short title for the diagnostic (at path/to/file.py:line:column)
+           |
+        42 | def foo(x: blah) -> None:
+           |            ^^^^ Span label
+           |
+        55 |     x = bar() + baz
+           |         ----- Sub-diagnostic label
+
+        Longer message describing the error.
+
+        note: Sub-diagnostic message without span
+        ```
+        """
+        if diag.span is None:
+            # Omit the title if we don't have a span, but a long message. This case
+            # should be fairly rare.
+            msg = diag.rendered_long_message or diag.rendered_title
+            self.buffer += textwrap.wrap(
+                f"{self.level_str(diag.level)}: {msg}",
+                self.MAX_MESSAGE_LINE_LEN,
+            )
```

Comment:
> Is the assumption that there won't be any if the parent doesn't have a span?

I think that's an ok assumption to make for now since basically 99% of diagnostics are going to come with a span. We can come back to that if it turns into an issue later.

I'll add a check though 👍